File: komorebi_desktop/src/api/feedback.ts
import { invoke } from "@tauri-apps/api/core";
import type { FeedbackStats } from "./types";

// ============================================================================
//  Phase 1: feedback telemetry
// ============================================================================

export async function feedbackRecord(args: {
  modelLabel: string;
  route: string;
  prompt: string;
  response: string;
  rating: 1 | -1;
  lang: string;
}): Promise<number> {
  return invoke<number>("feedback_record", {
    modelLabel: args.modelLabel,
    route: args.route,
    prompt: args.prompt,
    response: args.response,
    rating: args.rating,
    lang: args.lang,
  });
}

export async function feedbackStats(): Promise<FeedbackStats> {
  return invoke<FeedbackStats>("feedback_stats");
}

export async function feedbackPurge(): Promise<number> {
  return invoke<number>("feedback_purge");
}

export async function setTelemetryEnabled(enabled: boolean): Promise<void> {
  await invoke("set_telemetry_enabled", { enabled });
}

export async function setTelemetryEndpoint(url: string): Promise<void> {
  await invoke("set_telemetry_endpoint", { url });
}

// ============================================================================
//  Phase 2 stub: local LoRA training settings
// ============================================================================

export async function setTrainingEnabled(enabled: boolean): Promise<void> {
  await invoke("set_training_enabled", { enabled });
}

export async function setTrainingMaxCpuPct(pct: number): Promise<void> {
  await invoke("set_training_max_cpu_pct", { pct });
}

export async function setTrainingBatteryFloorPct(pct: number): Promise<void> {
  await invoke("set_training_battery_floor_pct", { pct });
}

export async function setTrainingMinExamples(n: number): Promise<void> {
  await invoke("set_training_min_examples", { n });
}

export async function setTrainingSchedule(schedule: string): Promise<void> {
  await invoke("set_training_schedule", { schedule });
}

File: dix_desktop/src-tauri/src/commands/feedback.rs
//! Phase 1: feedback telemetry commands.

use crate::settings;
use tauri::{AppHandle, Wry};

#[derive(serde::Serialize)]
pub struct FeedbackStatsDto {
    pub pending: i64,
    pub uploaded: i64,
    pub telemetry_enabled: bool,
    pub anon_token: Option<String>,
}

#[tauri::command]
pub fn feedback_record(
    app: AppHandle<Wry>,
    model_label: String,
    route: String,
    prompt: String,
    response: String,
    rating: i32,
    lang: String,
) -> Result<i64, String> {
    crate::feedback::record(
        &app,
        &model_label,
        &route,
        &prompt,
        &response,
        rating,
        &lang,
    )
}

#[tauri::command]
pub fn feedback_stats(app: AppHandle<Wry>) -> Result<FeedbackStatsDto, String> {
    let s = crate::feedback::stats(&app)?;
    Ok(FeedbackStatsDto {
        pending: s.pending,
        uploaded: s.uploaded,
        telemetry_enabled: settings::get_telemetry_enabled(&app),
        anon_token: settings::get_anon_token(&app),
    })
}

#[tauri::command]
pub fn feedback_purge(app: AppHandle<Wry>) -> Result<i64, String> {
    let n = crate::feedback::purge(&app)?;
    // Rotate the anonymous token alongside a purge so the user gets a
    // clean slate (the previous token can no longer be linked to any
    // future contributions).
    let _ = settings::rotate_anon_token(&app);
    Ok(n)
}

#[tauri::command]
pub fn set_telemetry_enabled(app: AppHandle<Wry>, enabled: bool) -> Result<(), String> {
    settings::set_telemetry_enabled(&app, enabled).map_err(|e| e.to_string())?;
    if enabled {
        // Ensure a token exists so the first upload doesn't have to wait.
        let _ = settings::ensure_anon_token(&app);
    }
    Ok(())
}

#[tauri::command]
pub fn set_telemetry_endpoint(app: AppHandle<Wry>, url: String) -> Result<(), String> {
    settings::set_telemetry_endpoint(&app, &url).map_err(|e| e.to_string())
}
